File: server/apps/nexotype/schemas/commercial/transaction_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import date, datetime

# ==========================================
# Request Schemas
# ==========================================

class TransactionCreate(BaseModel):
    """Schema for creating a new transaction"""
    buyer_id: int
    seller_id: int | None = None
    asset_id: int | None = None
    patent_id: int | None = None
    transaction_type: str
    value_usd: float | None = None
    announced_date: date


class TransactionUpdate(BaseModel):
    """Schema for updating a transaction"""
    buyer_id: int = None
    seller_id: int | None = None
    asset_id: int | None = None
    patent_id: int | None = None
    transaction_type: str = None
    value_usd: float | None = None
    announced_date: date = None


# ==========================================
# Response Schemas
# ==========================================

class TransactionDetail(BaseModel):
    """Schema for transaction details"""
    id: int
    buyer_id: int
    seller_id: int | None = None
    asset_id: int | None = None
    patent_id: int | None = None
    transaction_type: str
    value_usd: float | None = None
    announced_date: date
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class TransactionListResponse(BaseModel):
    """Response schema for listing transactions"""
    success: bool
    data: list[TransactionDetail] | None = None
    count: int = 0
    error: str | None = None


class TransactionResponse(BaseModel):
    """Response schema for single transaction operations"""
    success: bool
    data: TransactionDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/commercial/transaction.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const TransactionTypeEnum = z.enum(['Acquisition', 'Licensing', 'Divestiture']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** Transaction.transaction_type */
export const TRANSACTION_TYPE_OPTIONS = [
  { label: 'Acquisition', value: 'Acquisition' },
  { label: 'Licensing', value: 'Licensing' },
  { label: 'Divestiture', value: 'Divestiture' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const TransactionSchema = z.object({
  id: z.number(),
  buyer_id: z.number().int(),
  seller_id: z.number().int().nullable().optional(),
  asset_id: z.number().int().nullable().optional(),
  patent_id: z.number().int().nullable().optional(),
  transaction_type: TransactionTypeEnum,
  value_usd: z.number().nullable().optional(),
  announced_date: z.string(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateTransactionSchema = z.object({
  buyer_id: z.number().int(),
  seller_id: z.number().int().nullable().optional(),
  asset_id: z.number().int().nullable().optional(),
  patent_id: z.number().int().nullable().optional(),
  transaction_type: TransactionTypeEnum,
  value_usd: z.number().nullable().optional(),
  announced_date: z.string().min(1),
});

export const UpdateTransactionSchema = z.object({
  buyer_id: z.number().int().optional(),
  seller_id: z.number().int().nullable().optional(),
  asset_id: z.number().int().nullable().optional(),
  patent_id: z.number().int().nullable().optional(),
  transaction_type: TransactionTypeEnum.optional(),
  value_usd: z.number().nullable().optional(),
  announced_date: z.string().min(1).optional(),
});

// ==========================================
// Types
// ==========================================

export type Transaction = z.infer<typeof TransactionSchema>;
export type CreateTransaction = z.infer<typeof CreateTransactionSchema>;
export type UpdateTransaction = z.infer<typeof UpdateTransactionSchema>;
export type TransactionType = z.infer<typeof TransactionTypeEnum>;

// ==========================================
// Response Types
// ==========================================

export type TransactionResponse = {
  success: boolean;
  data?: Transaction;
  error?: string;
};

export type TransactionsResponse = {
  success: boolean;
  data?: Transaction[];
  count?: number;
  error?: string;
};
